### PR TITLE
Karla Grajales | Sprint-2 - Module-Structuring-and-Testing-Data | SPRINT 2

### PR DESCRIPTION
**Pull request by Grajales-K**
## Learners, PR Template

Self checklist

- [x]  I have committed my files one by one, on purpose, and for a reason
- [x]  I have titled my PR with COHORT_NAME | FIRST_NAME LAST_NAME | REPO_NAME | WEEK 
- [x]  I have tested my changes
- [x]  My changes follow the [style guide](https://curriculum.codeyourfuture.io/guides/contributing/)
- [x]  My changes meet the [requirements](./README.md) of this task

## Changelist

In this project, I learned how to use string methods by referring to the MDN documentation, how to set and manage variables, and how data flows within functions. I also learned how to invoke functions properly, debug issues to understand why something doesn't work, perform operations on strings, and iterate through strings based on their length.

## Questions

Ask any questions you have for your reviewer.

### CODE REVIEW

**Review by cjyuan — COMMENTED** [review]
Code is good. 
Some of the explanations given in the `debug` and `errors` folders are not quite correct.

File: Sprint-2/debug/2.js [hunk 1]
```diff
@@ -1,11 +1,24 @@
...
 
+/*
+we can not run this function because we settle the variable with const and this won't allow us to reassign 
+the new value which we are trying to test in the console.log()
+
 const num = 103;
 
 function getLastDigit() {
   return num.toString().slice(-1);
 }
```
[the remark this answers is not in the document]
Comment (Grajales-K) [reply Sprint-2/debug/2.js]: Hi @cjyuan, 

Thank you for your review. When I run the code above, it displays the message: "TypeError: Assignment to constant variable." This is what I was trying to explain. I can only use this variable once in the function, so I need to test it with different values. To do this, I must declare it with `let`, which allows me to reassign the value. I update my prediction to this exercise. Thank you. commit [e2929c210dbc6170a15da1e14b8878052a6fad48](e2929c210dbc6170a15da1e14b8878052a6fad48)

File: Sprint-2/errors/0.js [hunk 2]
```diff
@@ -1,9 +1,37 @@
...
+  we are creating a variable with (let = nameVariable) but the same time we can not reassign
+  the value in the same line.
+  instead we should call directly the (str = valueAssigned) and with this will Work
+*/
+
+
+// function capitalize(str) {
+//   let str = `${str[0].toUpperCase()}${str.slice(1)}`;
+//   return str;
+// }
```
[the remark this answers is not in the document]
Comment (Grajales-K) [reply Sprint-2/errors/0.js]: Thanks for this @cjyuan , yes we have "SyntaxError because the var str has been declared " so to fix this we just assign the value. I update my response. [commit](e2929c210dbc6170a15da1e14b8878052a6fad48)

File: Sprint-2/implement/cases.js [hunk 3]
```diff
@@ -13,3 +13,62 @@
...
+
+// ================ Last letter capitalized  ===================
+
+let quote = 'He said, "It\'s a sunny day!"';
+
+function lastLetterCapitalized(str) {
+  let result = ""; 
+
+  for (let i = 0; i < str.length; i++) {
+    if (str[i + 1] === " " || i === str.length - 1) {
```
[the remark this answers is not in the document]
Comment (Grajales-K) [reply Sprint-2/implement/cases.js]: Thanks for this @cjyuan, It's still confusing but I updated it as you suggested. thank you. [commit](21765d4f1b54c97ad12d326fcb622e071877ea46)

File: Sprint-2/implement/to-pounds.js [hunk 4]
```diff
@@ -4,3 +4,79 @@
...
+
+// ==================== function handling decimal inputs  ========================
+
+priceAmounts = "5.23p";
+
+function convertPenceToPound(priceAmounts) {
+  const cleanInput = priceAmounts.substring(0, priceAmounts.length - 1);
+
+  if (cleanInput.includes(".")) {
+    const [pounds, pence] = cleanInput.split("."); //destructuring;
```
[the remark this answers is not in the document]
Comment (Grajales-K) [reply Sprint-2/implement/to-pounds.js:62]: no is not, y changed to "0.23p" thank you.  [commit](da212f6ea4b3a70c164fc24d47dca1f9dcf29e3c)

File: Sprint-2/debug/2.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (cjyuan) [reply Sprint-2/debug/2.js]: You had that error because you had this in your code (I didn't notice them before):
```
const num = 103;
num = 24                    // <-------- This cause the runtime error
```

The reason why the original code does not work is because:
```
const num = 103;

function getLastDigit() {
  // 1) There is no parameter declared in this function.
  // 2) There is no local "num" found inside this function, so the global "num" is used instead.
  return num.toString().slice(-1);    // The return value will always be 3
}

console.log(`The last digit of 42 is ${getLastDigit(42)}`);
console.log(`The last digit of 105 is ${getLastDigit(105)}`);
```

File: Sprint-2/implement/cases.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (cjyuan) [reply Sprint-2/implement/cases.js]: Which part do you still find confusing? It won't help you learn if you just update the code without full understanding.

File: Sprint-2/implement/to-pounds.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (cjyuan) [reply Sprint-2/implement/to-pounds.js:62]: I suppose you just want some extra practice. The smallest unit of pence is 1p. Any fraction of a pence would probably be rounded to the nearest pence.

**Comment by cjyuan** [comment]
Changed code is good. I left you additional comments in the resolved conversations.

File: Sprint-2/debug/2.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (Grajales-K) [reply Sprint-2/debug/2.js]: Yes, the reason was that I was testing again and did not notice that I had left. I removed it again; thank you.

File: Sprint-2/implement/cases.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (Grajales-K) [reply Sprint-2/implement/cases.js]: HI @cjyuan  is the way that the conditions, I asked to chatgpt but still don't get where can cause the error "potential issue with accessing out-of-bounds elements in the string when i becomes equal to str.length "

File: Sprint-2/implement/to-pounds.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (Grajales-K) [reply Sprint-2/implement/to-pounds.js:62]: @cjyuan  Yes I need to look into and fix this too. for now, I handle the dot. Now I need to understand why 5.23 is not taken as 500 pence and 23 pence. 
Thank you.

File: Sprint-2/implement/cases.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (cjyuan) [reply Sprint-2/implement/cases.js]: The index to the last character in `str` is `str.length-1`, so `str.length` or `-1` would be consider an out-of-bounds index.

In JS, `str[str.length]` happened to evaluate to `undefined` (without causing runtime error).
So your code can still work normally. 

Suppose the problem were "do something to the current character if the next character is not a space" and we wrote the if condition as 
```
  if (str[i+1] != ' ' || i == str.length - 1) {
    // do something here to str[i] 
  }
```
then the code would have "done something" to the last character even though there is no character after the last character.

---

**Modifying** an out-of-bounds array element changes the size of an array in JS.
```
  let A = [1,2];
  A[100] = 3;
  console.log(A.length);   // output 101
```

In some programming languages, program can crash when we access a string or array with an out-of-bounds index.

**In short, it is a bad practice to not paying attention to potential out-of-bounds index.**
